main/distributed-programming.md

```diff
@@ -142,3 +142,78 @@ corresponding methods. It performs the communication asynchronously. Method
 calls can take arguments that are objects in the current vat or presences for
 objects in other vats.
 
+### Notifiers
+
+DApps and other tools may want to be notified about changes to the state of a Zoe
+contract or a particular offer. We support this with a Notifier tool based on Promises,
+which allows many parties to subscribe to be notified without requiring that the
+publisher track a subscription list.
+
+This is currently supported by Zoe itself, which publishes updates to the state of
+offers (reallocations and completions), and by some contracts, which can publish
+current prices or other details specific to the contract.
+
+#### Getting Notifications
+
+Zoe has a public method `getOfferNotifier()`, and contracts will have similar
+methods. This provides a long-lived notifier object associated with a particular
+stream of updates. At any time, you can call `notifier.getUpdateSince()`, and it will
+return a record `{ value, updateHandle, done }`. `value` is the current state,
+according to the source. `done` will be false until the stream reaches a final state,
+after which `value` will never change, and the same record will always be
+returned. `updateHandle` allows you to ask to be notified the next time there's a
+change to the state.
+
+If you call `getUpdateSince(oldUpdateHandle)` with no handle, or any `updateHandle`
+other than the most recent one, the notifier will immediately return a record with
+the current state. If you call `getUpdateSince(updateHandle)` with the most-recently
+issued `updateHandle`, you'll get a promise for the next record, which will be
+resolved the next time there's a change in the state.
+
+This approach to notification doesn't give any way to get a list of all state
```

Comment:
There are 3 cases, and they are largely based on the model rather than the client:
1. the model needs all deltas
2. the model needs only the last state
3 the model needs the last state but it's big and deltas may be better

For 1, the deltas should just be  part  of the model's state, and not somehow magic in the notification architecture; e.g., the edit history for multi-master editing and undos.

For 2, the value is just the lastest state.

For 3, we have a couple of options, and the fact that we provide a record is interesting. First, all this is out of scope for our current work! :)  The model can keep as much memory of the last N edits as it wants, associated with the update handle for them. Any particular notifier update can provide either the full state or a description of the set of deltas from the provided `updateHandle`. This could be done by an app (e.g., pixel demo) by open-coding it's own notifier, but could  someday be supported generically. Because we pass a record, adding an alternative `deltas` keyword instead of `value` is an option.  Details TBD :).

Reply:
Done.